src/librustc_resolve/build_reduced_graph.rs

```diff
@@ -584,7 +588,7 @@ impl<'a, 'b> BuildReducedGraphVisitor<'a, 'b> {
         let parent_scope = &self.parent_scope;
         let parent = parent_scope.module;
         let expansion = parent_scope.expansion;
-        let ident = item.ident.gensym_if_underscore();
```

Comment:
Hmm, I'd expect every `_` name to get `SyntaxContext` from a fresh macro expansion with def-site in `empty_module`, like it was done for other gensyms in https://github.com/rust-lang/rust/pull/63919.

Thus we'll reuse the same common mechanism, and when hygiene data becomes encodable we'll be able to properly encode `_`s in metadata as well, thus resolving the "FIXME: We shouldn't create the gensym here" below.

Reply:
Glob imports operate on the actual hygiene data: https://github.com/rust-lang/rust/blob/da01d31ff97feb199036359c364aedc809711123/src/librustc_resolve/resolve_imports.rs#L563

I'll see if there's a way to do this without increasing the complexity of `reverse_glob_adjust`

Reply:
Ah, right, the difference is that `_`s bring traits into scope for the purpose of type-based resolution, and names from opaque macros don't.

This difference is orthogonal to globs, so perhaps it's collection of traits in scope that needs to be adjusted, not `reverse_glob_adjust`.

(It's somewhat unfortunate that the difference exists and the features cannot be fully unified, but in both cases the behavior looks entirely reasonable and appropriate.)

Reply:
I'm specifically concerned about preserving the behavior of the following:

```rust
mod x {
    use Trait as _;
}
use x::*;
fn f() {
    ().method_on_trait(); 
    // ^This resolves if the whole snippet comes from a single macro expansion
    // It generally doesn't resolve otherwise
}
```

I think I have a working solution for this, which I'll push once I'm happy.